File: NativeCanvas/ios/Canvas/CanvasView.m
//
//  CanvasView.m
//  graphic
//
//  Created by sam on 2018/11/1.
//  Copyright © 2018 sam. All rights reserved.
//

#import "CanvasView.h"


@implementation CanvasView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _actions = [[NSMutableArray alloc] initWithCapacity:0];
    }
    return self;
}

- (void)initWithContext:(CGContextRef)context
{
    if (!_context) {
        _context = [CanvasRenderingContext2D new];
    }
    _context.context = context;
}

- (void)invalidate
{
    [self setNeedsDisplay];
}

- (void)clear
{
    [self.context initOrResetProperty];
}

- (void)drawRect:(CGRect)rect
{
    [self initWithContext:UIGraphicsGetCurrentContext()];
    [self runActions];
    [self clear];
}

/**
 * 未知参数的反射：
 * https://blog.csdn.net/imanapple/article/details/50477148
 * https://github.com/chukong/quick-cocos2d-x/blob/master/lib/cocos2d-x/scripting/lua/cocos2dx_support/platform/ios/CCLuaObjcBridge.mm (156 line)
 */
- (void)callMethodWithArguments:object method:(NSString *)method arguments:(NSArray *)arguments
{
    // @TODO 将 invocation 存起来
    SEL selector = NSSelectorFromString(method);
    NSMethodSignature *signature = [object methodSignatureForSelector:selector];

    if (signature.numberOfArguments == 0) {
        RCTLogWarn(@"callMethodWithArguments:%@ no selector", method);
        return;
    }
    if (signature.numberOfArguments > [arguments count] + 2) {
        RCTLogWarn(@"callMethodWithArguments:%@ not enough aruments", method);
        return;
    }

    NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
    [invocation setTarget:object];
    [invocation setSelector:selector];

    for (int i = 0; i < [arguments count]; i++) {
        NSObject *arg = [arguments objectAtIndex:i];
        // 类型转换
        if ([arg isKindOfClass:[NSNull class]]) {
            arg = nil;
        } else if ([arg isKindOfClass:[NSNumber class]]) {
            NSNumber *number = (NSNumber *)arg;
            const char *numberType = [number objCType];
            if ((strcmp(numberType, @encode(BOOL)) == 0) || [arg isKindOfClass:[@YES class]]) {
                bool value = [(NSNumber *)arg boolValue];
                [invocation setArgument:&value atIndex:2 + i];
            } else if (strcmp(numberType, @encode(int)) == 0) {
                int value = [(NSNumber *)arg intValue];
                [invocation setArgument:&value atIndex:2 + i];
            } else {
                CGFloat value = [(NSNumber *)arg floatValue];
                [invocation setArgument:&value atIndex:2 + i];
            }
            continue;
        }
        [invocation setArgument:&arg atIndex:2 + i];
    }

    [invocation invoke]; // Invoke the selector
}

/**
 * @{
 *   @"method" : @"",
 *   @"arguments" : @[]
 * };
 */
- (void)runActions
{
    for (NSDictionary *action in _actions) {
        NSString *method = [action objectForKey:@"method"];
        [self callMethodWithArguments:_context method:method arguments:[action objectForKey:@"arguments"]];
    }
}

@end

File: android/Graphic/app/src/main/java/com/example/sam/graphic/canvas/CanvasView.java
package com.example.sam.graphic.canvas;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.view.View;

import java.util.ArrayList;
import java.lang.Object;
import java.util.HashMap;

/**
 * Created by sam on 2018/11/21.
 */

public class CanvasView extends View {
  private static final JavaModuleWrapper module = new JavaModuleWrapper(CanvasRenderingContext2D.class);

  ArrayList<HashMap> actions = new ArrayList<>();
  private final CanvasRenderingContext2D context = new CanvasRenderingContext2D();

  public CanvasView(Context context) {
    super(context);
  }

  /**
   * { "method": "", "arguments": [] };
   */
  private void runAction(String method, Object[] arguments) {
    module.invoke(context, method, arguments);
  }

  private void runActions() {
    for (HashMap action : actions) {
      runAction((String) action.get("method"), (Object[]) action.get("arguments"));
    }
  }

  public void setActions(ArrayList<HashMap> drawActions){
    actions = new ArrayList<HashMap>(drawActions);
  }

  @Override
  protected void onDraw(Canvas canvas) {
    super.onDraw(canvas);

    if (context.getCanvas() == null) {
      context.setCanvas(canvas);
    }

    runActions();
  }
}
